src/Framework/Models/Model.php

```diff
@@ -0,0 +1,337 @@
+    {
+        if (is_null($value)) {
+            return true;
+        }
+
+        $type = $this->getPropertyType($key);
+
+        switch ($type) {
+            case 'int':
+                return is_int($value);
+            case 'string':
+                return is_string($value);
+            case 'bool':
+                return is_bool($value);
+            case 'array':
+                return is_array($value);
+            default:
+                return $value instanceof $type;
+        }
+    }
+
+    /**
+     * Validate property type
+     *
+     * @param  string  $key
+     * @param  mixed  $value
+     *
+     * @return void
+     *
+     * @throws InvalidArgumentException
+     */
+    protected function validatePropertyType($key, $value)
+    {
+        if (!$this->isPropertyTypeValid($key, $value)) {
+            $type = $this->getPropertyType($key);
+
+            throw new InvalidArgumentException("Invalid attribute assignment. '$key' should be of type: '$type'");
+        }
+    }
+
+    /**
+     * Get the property type
+     *
+     * @unreleased
+     *
+     * @param $key
+     * @return string
+     */
+    protected function getPropertyType($key)
+    {
+        return strtolower(trim($this->properties[$key]));
+    }
+
+    /**
+     * @unreleased
+     *
+     * @return array
+     */
+    public function toArray()
+    {
+        return $this->attributes;
+    }
+
+    /**
+     * @unreleased
+     *
+     * @return array
+     */
+    public function getAttributes()
+    {
+        return $this->attributes;
+    }
+
+    /**
+     * @return int[]|string[]
+     */
+    public static function propertyKeys()
+    {
+        return array_keys((new static)->properties);
+    }
+
+    /**
+     * Dynamically retrieve attributes on the model.
+     *
+     * @unreleased
+     *
+     * @param  string  $key
+     * @return mixed
+     */
+    public function __get($key)
+    {
+        if (array_key_exists($key, $this->relationships)) {
+            return $this->getRelationship($key);
+        }
+
+        return $this->getAttribute($key);
+    }
+
+    /**
+     * Dynamically set attributes on the model.
+     *
+     * @unreleased
+     *
+     * @param  string  $key
+     * @param  mixed  $value
+     * @return void
+     */
+    public function __set($key, $value)
+    {
+        $this->setAttribute($key, $value);
+    }
+
+    /**
+     * Determine if an attribute exists on the model.
+     *
+     * @unreleased
+     *
+     * @param  string  $key
+     * @return bool
+     */
+    public function __isset($key)
+    {
+        return isset($this->attributes[$key]);
+    }
+
+    /**
+     * @unreleased
+     *
+     * @param $key
+     *
+     * @return Model|Model[]
+     *
+     * @throws InvalidArgumentException
+     */
+    protected function getRelationship($key)
+    {
+        if (!is_callable([$this, $key])) {
+            throw new InvalidArgumentException("$key() does not exist in.");
```

Comment:
"does not exist in..." — is this missing the class or something?

Reply:
woops no just does not exist is good 73aa2c68